File: workers/scalacli/worker.scala
import scala.io.StdIn

object Worker {
  private def send(message: String): Unit = {
    println(message)
    Console.flush()
  }

  private def extract(line: String, marker: String): String = {
    val idx = line.indexOf(marker)
    if (idx < 0) "" else {
      val rest = line.substring(idx + marker.length)
      val end = rest.indexOf('"')
      if (end < 0) "" else rest.substring(0, end)
    }
  }

  def main(args: Array[String]): Unit = {
    Iterator.continually(StdIn.readLine()).takeWhile(_ != null).foreach { line =>
      if (line.contains("\"type\":\"HELLO\"")) {
        send("""{"type":"REGISTER","protocol":"ofp/1","workerId":"scala-cli-lines-01","language":"scalacli","runtimeVersion":"1.15.0","workerVersion":"0.1.0","capabilities":[{"name":"text.lines-scala"}]}""")
      } else if (line.contains("\"type\":\"JOB_START\"")) {
        if (!line.contains("\"capability\":\"text.lines-scala\"")) {
          send("""{"type":"JOB_ERROR","jobId":"unknown","error":"unsupported capability"}""")
        } else {
          val jobId = {
            val value = extract(line, "\"jobId\":\"")
            if (value.isEmpty) "job-unknown" else value
          }
          val text = extract(line, "\"text\":\"")
          val lineCount =
            if (text.isEmpty) 0
            else text.split("\\\\n", -1).length
          send(s"""{"type":"JOB_RESULT","jobId":"$jobId","output":{"lines":$lineCount}}""")
        }
      } else if (line.contains("\"type\":\"SHUTDOWN\"")) {
        sys.exit(0)
      }
    }
  }
}

File: workers/octave/worker.m
function worker()
  while (true)
    line = fgetl(stdin);
    if (!ischar(line))
      break;
    endif

    if (!isempty(strfind(line, '"type":"HELLO"')))
      send_message('{"type":"REGISTER","protocol":"ofp/1","workerId":"octave-mean-01","language":"octave","runtimeVersion":"11.3.0","workerVersion":"0.1.0","capabilities":[{"name":"math.mean"}]}');
    elseif (!isempty(strfind(line, '"type":"JOB_START"')))
      if (isempty(strfind(line, '"capability":"math.mean"')))
        send_message('{"type":"JOB_ERROR","jobId":"unknown","error":"unsupported capability"}');
      else
        job_id = extract_value(line, '"jobId":"');
        if (isempty(job_id))
          job_id = 'job-unknown';
        endif
        values = extract_numbers(line);
        count = numel(values);
        avg = 0;
        if (count > 0)
          avg = mean(values);
        endif
        send_message(sprintf('{"type":"JOB_RESULT","jobId":"%s","output":{"count":%d,"mean":%g}}', job_id, count, avg));
      endif
    elseif (!isempty(strfind(line, '"type":"SHUTDOWN"')))
      break;
    endif
  endwhile
endfunction

function value = extract_value(line, marker)
  value = '';
  start_pos = strfind(line, marker);
  if (isempty(start_pos))
    return;
  endif

  from = start_pos(1) + length(marker);
  if (from > length(line))
    return;
  endif

  rest = line(from:end);
  stop_pos = strfind(rest, '"');
  if (isempty(stop_pos))
    return;
  endif

  value = rest(1:(stop_pos(1) - 1));
endfunction

function values = extract_numbers(line)
  values = [];
  marker = '"numbers":[';
  start_pos = strfind(line, marker);
  if (isempty(start_pos))
    return;
  endif

  from = start_pos(1) + length(marker);
  if (from > length(line))
    return;
  endif

  rest = line(from:end);
  end_pos = strfind(rest, ']');
  if (isempty(end_pos))
    return;
  endif

  body = rest(1:(end_pos(1) - 1));
  if (isempty(body))
    return;
  endif

  parts = strsplit(body, ',');
  for i = 1:numel(parts)
    piece = strtrim(parts{i});
    if (!isempty(piece))
      values(end + 1) = str2double(piece);
    endif
  endfor
endfunction

function send_message(text)
  fprintf('%s\n', text);
  fflush(stdout);
endfunction

worker();
